anilist/src/test/java/com/huchihaitachi/anilist/AnilistPresenterTest.kt

```diff
@@ -1,80 +1,342 @@
-  @MockK private lateinit var getStringResourceUseCase: GetStringResourceUseCase
-  @MockK private lateinit var anilistView: AnilistView
-  @MockK private lateinit var  rxSchedulers: RxSchedulers
+  @MockK private lateinit var loadPage: LoadPageUseCase
+  @MockK private lateinit var refreshPage: RefreshPageUseCase
+  @MockK private lateinit var loadAnime: LoadAnimeUseCase
+  @MockK private lateinit var getStringResource: GetStringResourceUseCase
+  @MockK private lateinit var view: AnilistView
+  @MockK private lateinit var rxSchedulers: RxSchedulers
+  private lateinit var anime: List<Anime>
+  private lateinit var trampoline: Scheduler
+  private lateinit var testScheduler: TestScheduler
 
   @Before
   fun setup() {
     MockKAnnotations.init(this)
-    mockIntents()
+    setupView()
+    anime = listOf(Anime(1), Anime(2), Anime(3), Anime(4), Anime(5), Anime(6))
   }
 
   @Test
   fun `load first page test`() {
-    val initialState = AnilistViewState()
-    val anilistPresenter = AnilistPresenter(
-      loadPageUseCase,
-      loadAnimeUseCase,
-      getStringResourceUseCase,
-      AnilistViewState(),
-      rxSchedulers
-    )
-    val anime = listOf(
-      Anime(1), Anime(2), Anime(3), Anime(4), Anime(1), Anime(2), Anime(3), Anime(4)
-    )
-    val testScheduler = TestScheduler()
-    every { rxSchedulers.io } returns testScheduler
-    every { rxSchedulers.ui } returns testScheduler
-    anilistPresenter.bind(anilistView)
-    val loadAnimePageMock: PublishSubject<Unit> = PublishSubject.create()
-    every { anilistView.loadAnimePage } returns loadAnimePageMock
-    every { loadPageUseCase(any(), any()) } returns Single.just(Page(8, 1, true, anime))
-    every { anilistView.render(any()) } returns testScheduler.advanceTimeBy(100L, MILLISECONDS)
-    val expectedResult = AnilistViewState(PAGE, null, PageState(anime, 1, true), null)
-    anilistPresenter.bindIntents()
-    testScheduler.advanceTimeBy(10000L, TimeUnit.MILLISECONDS)
-    loadAnimePageMock.onNext(Unit)
-    verifySequence {
-      anilistView.render(initialState)
-      anilistView.render(expectedResult)
+    setupTrampoline()
+    val initialState = AnilistViewState(pageState = PageState(null, 0, true))
+    val presenter = AnilistPresenter(
+      loadPage, refreshPage, loadAnime, getStringResource, initialState, rxSchedulers
+    )
+    val loadPageIntentMock: PublishSubject<Unit> = PublishSubject.create()
+    mockIntents(loadPageIntent = loadPageIntentMock)
+    every { loadPage(any(), any()) } returns Single.just(
+      Page(1, anime.size, true, anime, 0, 0)
+    )
+    presenter.bind(view)
+    presenter.bindIntents()
+    val expectedInitial = AnilistViewState(pageState = PageState(null, 0, true))
+    val expectedLoading = AnilistViewState(
+      loading = PAGE,
+      pageState = PageState(null, 0, true),
+    )
+    val expectedResult = AnilistViewState(pageState = PageState(anime, 1, true))
+    loadPageIntentMock.onNext(Unit)
```

Comment:
General case is testing only order of calls, not exact sequence, so states may not contain the previous state at all because some intermediate state was omitted. The question is: should I copy the previous state if possible or use a unified way of creating states (through the constructor). The second thing is the copy() function may work incorrectly which cause test fails (not in this case, but in general)

Reply:
Yea, `you should copy the previous state if possible`